File: backend-node/src/api/authEndpoint.js
const { applyToken, inquiryUserInfo } = require('../alipay');
const { createJWE } = require('../jwe/jwe');

function initAuthEndpoint(router) {
    router.post('/auth/apply-token', async (req, res) => {
        try {
            const { auth_code } = req.body;

            if (!auth_code) {
                return res.status(400).json({ error: 'auth_code is required' });
            }

            console.log('=================================================================');
            console.log('STARTING AUTH TOKEN EXCHANGE');
            console.log('=================================================================');

            const tokenResponse = await applyToken(auth_code);

            console.log('[SUCCESS] Token response received:');
            console.log(JSON.stringify(tokenResponse, null, 2));
            console.log('');

            if (tokenResponse.result.resultCode !== 'SUCCESS') {
                console.log(`[ERROR] Invalid token response: ${tokenResponse.result.resultMessage}`);
                return res.status(400).json({
                    error: 'Invalid token response: ' + tokenResponse.result.resultMessage
                });
            }

            const info = await inquiryUserInfo(tokenResponse.accessToken);

            console.log('[SUCCESS] User info retrieved:');
            console.log(JSON.stringify(info, null, 2));
            console.log('');

            const jweToken = await createJWE({
                user_id: info.userInfo.userId,
                access_token: tokenResponse.accessToken
            });

            console.log('[SUCCESS] Returning auth token to frontend');

            return res.json({
                token: jweToken
            });

        } catch (err) {
            console.log(`[ERROR] Auth endpoint error: ${err.message}`);
            return res.status(500).json({ error: err.message });
        }
    });
}

module.exports = { initAuthEndpoint };

File: backend-go/api/authEndpoint.go
package api

import (
	"encoding/json"
	"log"
	"superQiMiniAppBackend/alipay"
	"superQiMiniAppBackend/jwe"

	"github.com/gofiber/fiber/v2"
)

type authRequest struct {
	AuthCode string `json:"auth_code" validate:"required"`
}

func InitAuthEndpoint(group fiber.Router) {
	group.Post("/auth/apply-token", func(ctx *fiber.Ctx) error {
		var request authRequest
		if err := ctx.BodyParser(&request); err != nil {
			return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
		}

		log.Println("=================================================================")
		log.Println("STARTING AUTH TOKEN EXCHANGE")
		log.Println("=================================================================")

		tokenResponse, err := alipay.Interface.ApplyToken(request.AuthCode)
		if err != nil {
			log.Printf("[ERROR] Token exchange failed: %v\n", err)
			return fiber.NewError(fiber.StatusBadRequest, err.Error())
		}

		tokenResponseJson, _ := json.MarshalIndent(tokenResponse, "", "  ")
		log.Printf("[SUCCESS] Token response received:\n%s\n\n", string(tokenResponseJson))

		if tokenResponse.Result.ResultCode != "SUCCESS" {
			log.Printf("[ERROR] Invalid token response: %s\n", tokenResponse.Result.ResultMessage)
			return fiber.NewError(fiber.StatusBadRequest, "Invalid token response: "+tokenResponse.Result.ResultMessage)
		}

		log.Println("[INFO] Token exchange successful")
		log.Printf("[INFO] Customer ID: %s\n", tokenResponse.CustomerID)
		log.Println("[INFO] User/Merchant detailed info can be retrieved via separate endpoints")

		// Return a JWE to the MiniApp containing the access token and customer ID
		// The customerID is returned from the token exchange and can be either userId or merchantId
		jweToken, err := jwe.CreateJWE(jwe.TokenClaims{
			UserID:      tokenResponse.CustomerID,
			AccessToken: tokenResponse.AccessToken,
		})

		if err != nil {
			log.Printf("[ERROR] Failed to create JWE token: %v\n", err)
			return fiber.NewError(fiber.StatusUnauthorized, err.Error())
		}

		response := fiber.Map{
			"token": jweToken,
		}

		log.Println("[SUCCESS] Returning auth token to frontend")
		return ctx.JSON(response)
	})
}
